File: TypeScript/ExampleTests/common/NewPostTyped.ts
import { NewPostString } from "./NewPostString.js";
import * as _json from "./json.js";

export class NewPostTyped {
  title: string;
  body: string;
  userId: number;

  constructor(title: string, body: string, userId: number) {
    this.title = title;
    this.body = body;
    this.userId = userId;
  }

  static fromStringObj(s: NewPostString): NewPostTyped {
    return new NewPostTyped(
      s.title,
      s.body,
      s.userId !== "" ? Number(s.userId) : 0
    );
  }

  toStringObj(): NewPostString {
    return new NewPostString(
      String(this.title),
      String(this.body),
      String(this.userId)
    );
  }

  static toStringList(list: NewPostTyped[]): NewPostString[] {
    return list.map(t => t.toStringObj());
  }

  static fromStringList(list: NewPostString[]): NewPostTyped[] {
    return list.map(s => NewPostTyped.fromStringObj(s));
  }

  toJsonValue(): Record<string, unknown> {
    return {
      title: this.title,
      body: this.body,
      userId: this.userId,
    };
  }

  toJSON(): string { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m: unknown): NewPostTyped {
    return new NewPostTyped(
      _json.asString(_json.requireField(m, "title"), "title"),
      _json.asString(_json.requireField(m, "body"), "body"),
      _json.asInt(_json.requireField(m, "userId"), "userId")
    );
  }

  static fromJSON(text: string): NewPostTyped {
    return NewPostTyped.fromJsonValue(_json.parse(text));
  }

  static toJSONList(items: readonly NewPostTyped[]): string {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text: string): NewPostTyped[] {
    const raw = _json.asArray(_json.parse(text), "NewPostTyped") ?? [];
    return raw.map((e) => NewPostTyped.fromJsonValue(e));
  }

  toString(): string {
    return `title=${this.title}, body=${this.body}, userId=${this.userId}`;
  }

  equals(other: NewPostTyped): boolean {
    return this.title === other.title
      && this.body === other.body
      && this.userId === other.userId;
  }
}

File: Rust/ExampleTests/common/newpost_typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::newpost_string::NewPostString;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct NewPostTyped {
    pub title: String,
    pub body: String,
    pub userid: i32,
}

impl NewPostTyped {
    pub fn from_str_struct(s: &NewPostString) -> Self {
        Self {
            title: s.title.clone(),
            body: s.body.clone(),
            userid: s.userid.parse::<i32>().unwrap_or_default(),
        }
    }

    pub fn to_str_struct(&self) -> NewPostString {
        NewPostString {
            title: self.title.clone(),
            body: self.body.clone(),
            userid: self.userid.to_string(),
        }
    }

    pub fn to_string_list(list: &[NewPostTyped]) -> Vec<NewPostString> {
        list.iter().map(|t| t.to_str_struct()).collect()
    }

    pub fn from_string_list(list: &[NewPostString]) -> Vec<NewPostTyped> {
        list.iter().map(NewPostTyped::from_str_struct).collect()
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("title".to_string(), json::Value::Str(self.title.clone())),
            ("body".to_string(), json::Value::Str(self.body.clone())),
            ("userId".to_string(), json::Value::number_from_i64(self.userid as i64)),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            title: json::as_string(json::require(v, "title")?, "title")?,
            body: json::as_string(json::require(v, "body")?, "body")?,
            userid: json::as_i32(json::require(v, "userId")?, "userId")?,
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[NewPostTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<NewPostTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "NewPostTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}
